synapse/rest/client/receipts.py

```diff
@@ -109,6 +109,46 @@ async def on_POST(
 
         return 200, {}
 
+    async def _is_valid_thread_id(self, event_id: str, thread_id: str) -> bool:
+        """
+        The thread ID provided must relate (in a vague sense) to the event ID.
+
+        We check this to ensure clients aren't sending bogus receipts.
+
+        A thread ID is considered valid if:
+
+        1. The event has a thread relation which matches the thread ID.
+        2. The event has children events which form a thread relation (i.e. the
+           event is a thread root).
+        2. The event is related to an event (recursively) which satisfies 1 or 2.
```

Comment:
I think you want `3.` here.

Is any relation permitted here? There's potential to confuse this with "relate" on +114.

https://github.com/matrix-org/synapse/issues/14168#issuecomment-1277670830 talks about a limit to recursion that we go on to implement below and in the next commit. Should we mention that here?

Reply:
> Is any relation permitted here? There's potential to confuse this with "relate" on +114.

Yes (as long as it abides by the rules of that relation).

> https://github.com/matrix-org/synapse/issues/14168#issuecomment-1277670830 talks about a limit to recursion that we go on to implement below and in the next commit. Should we mention that here?

Maybe? It is somewhat an implementation detail, but likely worth mentioning. Maybe I'll leave it a bit vague and say "recursively, up to a point" or something?